%        File: documentation.tex
%     Created: Mo Nov 11 02:00  2013 C
% Last Change: Mo Nov 11 02:00  2013 C
%
\documentclass[a4paper]{article}

\usepackage{doi,hyperref,verbatimbox,verbatim,url}
\usepackage[numbers]{natbib}

\usepackage[Q=yes,bsdiv=yes]{examplep}
\usepackage{amsmath}

\newcommand{\plasma}[0]{\texttt{plasma}}
\newcommand{\discrover}[0]{\texttt{discrover}}
\newcommand{\dinuclshuffle}[0]{\texttt{discrover-shuffle}}

\newcommand{\version}[0]{@GIT_DESCRIPTION@}
\newcommand{\theurl}[0]{\url{https://github.com/maaskola/discrover}}
\newcommand{\TikZ}{Ti\emph{k}Z}
\newcommand{\ignore}[1]{}

% Motif topology illustration code
\usepackage{pgf}
\usepackage{tikz}

% \usetikzlibrary{backgrounds}
% \usetikzlibrary{shadows.blur}
% \usetikzlibrary{decorations.pathreplacing}
\usetikzlibrary{positioning}
\usetikzlibrary{calc}
% \usetikzlibrary{shapes}
% \usetikzlibrary{fit,shapes.geometric}

\newcommand{\minCircleSize}[0]{0.8cm}

\tikzset{%
  var/.style={draw, circle, thick, fill=gray!30, minimum size=\minCircleSize, inner sep=0},
  myvar/.style={var, minimum size=0.5cm},
  Var/.style={var, ellipse, align=center},
  dedge/.style={draw, thick, -latex},
  edge/.style={draw, thick},
  note/.style={rectangle callout, fill=mdcblue!25, line join=bevel, text=black},
  % factor/.style={rectangle, minimum size=0.15cm, minimum height=0.15cm, fill=black, inner sep=0pt}
  factor/.style={rectangle, fill=black, minimum size=2mm, inner sep=0}
}

\newcommand{\dist}[0]{0.35cm}

\newcommand{\Core}[0]{%
  \node[emit] at (0,0) (m1) {$M_1$};
  \node[emit, right=\dist of m1] (m2) {$M_2$};
  \node[emit, right=\dist of m2] (m3) {$M_3$};
  \node[emit, right=\dist of m3] (m4) {$M_4$};
  \node[emit, right=\dist of m4] (m5) {$M_5$};
  \node[emit, right=\dist of m5] (m6) {$M_6$};
  \node[emit, right=\dist of m6] (m7) {$M_7$};
  \node[emit, right=\dist of m7] (m8) {$M_8$};
  \coordinate (x1) at ($(m1)!0.5!(m2)$);
  \coordinate (x2) at ($(m2)!0.5!(m3)$);
  \coordinate (x3) at ($(m3)!0.5!(m4)$);
  \coordinate (x4) at ($(m4)!0.5!(m5)$);
  \coordinate (x5) at ($(m5)!0.5!(m6)$);
  \coordinate (x6) at ($(m6)!0.5!(m7)$);
  \coordinate (x7) at ($(m7)!0.5!(m8)$);

  \path[dedge] (m1) to (m2);
  \path[dedge] (m2) to (m3);
  \path[dedge] (m3) to (m4);
  \path[dedge] (m4) to (m5);
  \path[dedge] (m5) to (m6);
  \path[dedge] (m6) to (m7);
  \path[dedge] (m7) to (m8);
}

\newcommand{\SimpleCore}[0]{%
  \Core{}

  \node[emit, below=4*\dist of x4] (bg) {$B$};
  \path[dedge, loop above] (bg) to (bg);

  \path[dedge, bend left=20] (m8) to (bg);
  \path[dedge, bend left=20] (bg) to (m1);
}

\newcommand{\ExtendedCore}[0]{%
  \Core{}

  \node[emit, below=10*\dist of x4] (bg) {$B$};
  \path[dedge, loop above] (bg) to (bg);

  \node[delete, left=\dist of m1] (m0) {$M_0$};
  \node[delete, right=\dist of m8] (m9) {$M_9$};
  \path[dedge] (m0) to (m1);
  \path[dedge] (m8) to (m9);
  \path[dedge, bend left=40] (m9) to (bg);
  \path[dedge, bend left=40] (bg) to (m0);
}

\newcommand{\Insertions}[0]{%
  % Insertion states
  \node[emit, above=2*\dist of x1] (i1) {$I_1$};
  \node[emit, above=2*\dist of x2] (i2) {$I_2$};
  \node[emit, above=2*\dist of x3] (i3) {$I_3$};
  \node[emit, above=2*\dist of x4] (i4) {$I_4$};
  \node[emit, above=2*\dist of x5] (i5) {$I_5$};
  \node[emit, above=2*\dist of x6] (i6) {$I_6$};
  \node[emit, above=2*\dist of x7] (i7) {$I_7$};

  \path[dedge] (m1) to (i1);
  \path[dedge] (m2) to (i2);
  \path[dedge] (m3) to (i3);
  \path[dedge] (m4) to (i4);
  \path[dedge] (m5) to (i5);
  \path[dedge] (m6) to (i6);
  \path[dedge] (m7) to (i7);

  \path[dedge] (i1) to (m2);
  \path[dedge] (i2) to (m3);
  \path[dedge] (i3) to (m4);
  \path[dedge] (i4) to (m5);
  \path[dedge] (i5) to (m6);
  \path[dedge] (i6) to (m7);
  \path[dedge] (i7) to (m8);

}

\newcommand{\InsertionsSelf}[0]{%
  \path[dedge, loop above] (i1) to (i1);
  \path[dedge, loop above] (i2) to (i2);
  \path[dedge, loop above] (i3) to (i3);
  \path[dedge, loop above] (i4) to (i4);
  \path[dedge, loop above] (i5) to (i5);
  \path[dedge, loop above] (i6) to (i6);
  \path[dedge, loop above] (i7) to (i7);
}

\newcommand{\Deletions}[0]{%
  % Deletion states

  \node[delete, below=2*\dist of m1] (d1) {$D_1$};
  \node[delete, below=2*\dist of m2] (d2) {$D_2$};
  \node[delete, below=2*\dist of m3] (d3) {$D_3$};
  \node[delete, below=2*\dist of m4] (d4) {$D_4$};
  \node[delete, below=2*\dist of m5] (d5) {$D_5$};
  \node[delete, below=2*\dist of m6] (d6) {$D_6$};
  \node[delete, below=2*\dist of m7] (d7) {$D_7$};
  \node[delete, below=2*\dist of m8] (d8) {$D_8$};

  \path[dedge] (m0) to (d1);
  \path[dedge] (m1) to (d2);
  \path[dedge] (m2) to (d3);
  \path[dedge] (m3) to (d4);
  \path[dedge] (m4) to (d5);
  \path[dedge] (m5) to (d6);
  \path[dedge] (m6) to (d7);
  \path[dedge] (m7) to (d8);

  \path[dedge] (d1) to (d2);
  \path[dedge] (d2) to (d3);
  \path[dedge] (d3) to (d4);
  \path[dedge] (d4) to (d5);
  \path[dedge] (d5) to (d6);
  \path[dedge] (d6) to (d7);
  \path[dedge] (d7) to (d8);

  \path[dedge] (d1) to (m2);
  \path[dedge] (d2) to (m3);
  \path[dedge] (d3) to (m4);
  \path[dedge] (d4) to (m5);
  \path[dedge] (d5) to (m6);
  \path[dedge] (d6) to (m7);
  \path[dedge] (d7) to (m8);
  \path[dedge] (d8) to (m9);
}

\newcommand{\InsertionToDeletionEdges}[0]{%
  \path[dedge] (i1) to (d2);
  \path[dedge] (i2) to (d3);
  \path[dedge] (i3) to (d4);
  \path[dedge] (i4) to (d5);
  \path[dedge] (i5) to (d6);
  \path[dedge] (i6) to (d7);
  \path[dedge] (i7) to (d8);
}


\newcommand{\TikzEnv}[1]{%
  \begin{tikzpicture}[%
      genscan/.style={draw, thick, minimum size=\minCircleSize, inner sep=0},
      delete/.style={genscan, diamond},
      emit/.style={genscan, circle},
    ]
    #1
  \end{tikzpicture}
}
% End of motif topology illustration code

\hypersetup{%
  pdfauthor=Jonas Maaskola,
  pdftitle=Manual for discrover v\version{},
  pdfkeywords={Motif searching} {Discriminative learning} {Motif discovery} {Nucleic acids} {RNA} {DNA} {HMM} {Hidden Markov model},
  colorlinks=true     % false: boxed links; true: colored links
}

\title{Manual for \discrover{} v\version{}}
\author{Jonas Maaskola}
\date{\today}
\begin{document}
\maketitle

\begin{abstract}
  This is the manual of \discrover{} version \version{}.
  Please refer to the publication~\cite{Maaskola2014} for a full description of the method.

  The package provides programs for discriminative motif discovery of protein binding sites in nucleic acid sequences.
  It is available as free software under the GNU General Public License v3 from \theurl{}.

  This document briefly describes the purpose of this motif finding package and discusses important concepts.
  A tutorial explains the basic usage of the programs that are part of this package.
  Recipes are given that can be used to perform motif analysis for CLIP-Seq and ChIP-Seq data.
  Finally, further details of the program usage are described.
\end{abstract}

\tableofcontents

\section{Introduction}

Transcriptional and post-transcriptional regulation rely to a large extent on effective mechanisms that allow nucleic acid binding proteins to recognize specific sets of nucleic acids.
Aside from structural cues, binding of regulators is guided by sequence information (motifs) present in cognate nucleic acids.

Motif discovery is the problem of unraveling motifs recognized by a given nucleic acid binding protein from sequences known to harbor occurrences of the motif.
Classically, motif finding was marked by scarcity of data when only few sequences were available.
The introduction of microarray based technologies like ChIP-chip \cite{Ren2000,Iyer2001} and RIP-Chip \cite{Tenenbaum2000,Keene2006} made it possible to assay \textit{in vivo} sequence binding specificity on a genome-wide scale.
More recently sequencing based technologies, such as ChIP-Seq \cite{Robertson2007, Johnson2007} and CLIP-Seq \cite{Licatalosi2008,Sanford2009,Hafner2010} further increased the amounts of data yielded by single experiments and simultaneously improved the spatial resolution, reducing uncertainty about the exact location of \textit{in vivo} binding sites.
SELEX \cite{Ellington1990,Tuerk1990} and related sequencing based technologies \cite{Jolma2010}, and protein-binding microarrays (PBMs) \cite{Bulyk2001,Berger2006a} are targeted assays for the \textit{in vitro} sequence binding specificity of nucleic acid binding proteins.

Due to the central importance of the motif discovery problem in computational biology many algorithms addressing it have been developed over the last two decades \cite{Das2007}.
These algorithms employ  a variety of models for the sequence binding specificity of nucleic acid binding proteins, including discrete word-based models, as well as probabilistic models such as position weight matrices (PWM) \cite{Stormo1998}, and hidden Markov models (HMM) \cite{Rabiner1989}.
Word based approaches tend to be computationally efficient and allow for fast global optimization, but may fail for motifs that include weak positions \cite{Das2007}.
PWMs can be motivated from biophysical principles \cite{Berg1987,Stormo2000,Foat2006}.
General inference methods for HMMs offer a unified framework for biological sequence modeling \cite{Durbin1998}.
HMMs flexibly model both binding sites and their surrounding sequence context, may account for dependence of neighboring positions via higher order emissions, and length variability of motifs is idiomatically realized via non-constitutive states in the HMM topology.

Because of historically smaller data sizes, many of the presently available motif finding methods were not designed for data sets as large as those produced by current experiments.
Consequently, many of these abort or run unacceptably slow when applied to large data sets.
Thus, even after more than two decades of computational analysis of biological sequences, there is continued interest in the development of new analysis methods that leverage the full potential of large data sets.

\subsection{Overview}
Here we describe a discriminative learning method based on HMMs, realized and available as free software, to automatically discover binding site sequence motifs of nucleic acid binding proteins from arbitrary contrasts,
such as positive and negative example sequences.
Not all of the positive examples need contain binding site occurrences and not all negative examples need be devoid of them.
The framework is applicable to a broad variety of contrasts, comprising the comparison of strongly bound versus weakly bound targets, or of signal sequences with shuffled sequences.
It is also possible to discover context-dependent motifs, or to analyze data sets of different factors for mutually discriminative features.
When available, information from repeat experiments is leveraged by the method.

Putative motifs are identified by a measure of association that quantifies how strongly motif presence varies across conditions of a contrast.
For the chosen objective function, a discrete optimization phase determines IUPAC regular expressions as seeds from which HMMs are initialized.
Subsequently HMM parameters are improved with gradient optimization, which is demonstrated to robustly and effectively function with realistic data sizes.

Objective functions evaluated on multiple contrast may be combined into meta-objectives that allow for a joint analysis of repeat experiments, or to discover motifs that are differential in one contrast but not, or less so, in another.

\subsection{Objective functions}
Motifs are discovered by optimizing an objective function that tries to measure the amount of evidence for the motif.
There are many possible choices for objective functions for this purpose, and this package lets the user decide which objective function to use.

The classical objective function is likelihood.
The corresponding maximum likelihood (ML) principle dictates that parameters are to be chosen such that the probability for the observed data being generated by the statistical model with these parameters is to be as high as possible.

Most of the objective functions considered here are discriminative.
This is to say that they operate on multiple sets of sequences and try to identify motifs that are ``discriminative'' between these sets.
Here, the objective function of course defines exactly what is meant by the term ``discriminative''.

In particular the following objective functions are available:
\begin{itemize}
  \item Likelihood (ML)
  \item Classification posterior probability, also known as maximum mutual information estimation (MMIE)
  \item Difference of occurrence frequency (DFREQ)
  \item Likelihood difference (DLOGL)
  \item Matthews' correlation coefficient (MCC)
  \item Mutual information of condition and motif occurrence (MICO)
\end{itemize}
The publication describing this method, \cite{Maaskola2014}, presents experiments based on synthetic data that suggest that the objective functions MCC, MICO, and MMIE have higher motif discovery performance than the other objective functions analyzed.
Among these MICO is the most versatile, as it is compatible with the multiple motif discovery mode implemented in \discrover{}.
Hence, MICO is the recommended objective function.

\ignore{%
\subsection{Seeding}
To be written.
\subsection{Binding site HMM optimization}
To be written.
}

\section{Tutorial}
This tutorial demonstrates the basic usage of the motif discovery tools in this package.
Let us assume that you have one FASTA file with signal sequences, \verb|signal.fa|, and one with control sequences, \verb|control.fa|, and that you want to find motifs of length 8~nt that are discriminative between the two.
\subsection{Basic usage}
To perform a full analysis of discriminative motifs, \discrover{} automatically discovers the most discriminative IUPAC regular expression as seed, initializes an HMM on it, and subsequently optimizes the HMM parameters.
\subsubsection{Analysis of RNA-binding proteins}
To analyze RNA-binding proteins the following command already suffices:\\
\begin{verbbox}
discrover signal.fa control.fa -m 8
\end{verbbox}
\fbox{\theverbbox[t]}

\subsubsection{Analysis of DNA-binding proteins}
In case you want to analyze DNA-binding proteins, please add the \verb|-r| option or its long variant \verb|--revcomp| to also consider motif occurrences on the reverse complementary strand.\\
\begin{verbbox}
discrover signal.fa control.fa -m 8 -r
\end{verbbox}
\fbox{\theverbbox[t]}

\subsection{Manual seed discovery}
\label{tutorial:plasma}
Apart from the fully integrated analysis described above, it is also possible to manually run the seed discovery:\\
\begin{verbbox}
plasma signal.fa control.fa -m 8 -r
\end{verbbox}
\fbox{\theverbbox[t]}

Here, we are looking again for a DNA-binding protein's motif (due to the usage of the option \verb|-r|).
The command line interface of the two programs are intentionally designed to be similar, so that the most important command line options apply to both programs.

\subsubsection{Discovering seeds across a length range}
You can use the following motif specification to discover seeds from a given range of lengths.
Say you want to consider seeds of 5--10~nt length, then you can use this command:\\
\begin{verbbox}
plasma signal.fa control.fa -m 5-10 -r
\end{verbbox}
\fbox{\theverbbox[t]}

Note that the same syntax can also be used for automatic seed discovery with \discrover{}:
\\
\begin{verbbox}
discrover signal.fa control.fa -m 5-10 -r
\end{verbbox}
\fbox{\theverbbox[t]}



\subsection{Manual seed specification for \discrover{}}
Say that you have found a seed of interest for your protein, for example by manually using \plasma{} as described in section~\ref{tutorial:plasma}.
Let us assume it is the IUPAC regular expression \verb|tgtanatw|, where the \verb|n| stands for any nucleotide and the \verb|w| for either \verb|a| or \verb|t|.
You can now use this IUPAC regular expression as seed for an HMM and optimize the HMM parameters as follows:\\
\begin{verbbox}
discrover signal.fa control.fa -m tgtanatw
\end{verbbox}
\fbox{\theverbbox[t]}


\subsection{Finding multiple motifs}
For some applications it is interesting to discover multiple motifs in a dataset.
This may reveal motifs of co-factors of the assayed factor.
This is for example of relevance in analyzing ChIP-Seq data, where a given IP'ed transcription factor often acts in conjunction with other factors, whose binding sites are typically found nearby, and thus are also frequently enriched in ChIP-Seq sequences.

\discrover{} has two ways of dealing with multiple motifs, depending on whether or not the switch \verb|--multiple| is used, as described next.
In both modes, multiple seeds are first determined by giving a length range of interest as well as by looking for possibly multiple seeds of each length.
The general syntax for such a motif specification is:\\
\begin{verbbox}
-m 5-16x3
\end{verbbox}
\fbox{\theverbbox[t]}\\
Here, we use the three best seeds for each length of 5--16~nt.

\subsubsection{Multiple independent models with individual seeds}
When the switch \verb|--multiple| is not used, as in the next example, multiple seeds are independently sought, an HMM is initialized and independently optimized for each, and subsequently the single best HMM is reported.\\
\begin{verbbox}
discrover signal.fa control.fa -m 5-16x3 -r
\end{verbbox}
\fbox{\theverbbox[t]}

\subsubsection{HMMs of multiple motifs}
However when the switch \verb|--multiple| is used, as in the next example, a more refined procedure is used.
For the details of this mode please refer to the publication.
In particular, the supplementary material contains flowcharts illustrating how this mode builds models of multiple motifs.\\
\begin{verbbox}
discrover signal.fa control.fa -m 5-16x3 -r --multiple
\end{verbbox}
\fbox{\theverbbox[t]}

Different models are learned and tested when the switch \verb|--multiple| is used.
The final, accepted model's parameters and output files extend the label used for the file names with a \verb|.accepted| tag.

\subsection{Generating sequence logos for motifs}
The package provides multiple ways to generate sequence logos of motifs.
By default, \verb|discrover| will generate sequence logos in PNG and PDF format.
While \verb|plasma| does not by default generate sequence logos, the command line switches \verb|--png| and \verb|--pdf| respectively generate sequence logos in PNG and PDF format.

The same routines can also be used via a separate program: \verb|discrover-logo|.
It creates sequence logos from the \verb|.hmm| files generated by \verb|discrover|, from IUPAC regular expressions, or from motif matrices in files:\\
\begin{verbbox}
discrover-logo motif.hmm
\end{verbbox}
\fbox{\theverbbox[t]}\\
\begin{verbbox}
discrover-logo -i uguahata
\end{verbbox}
\fbox{\theverbbox[t]}\\
\begin{verbbox}
discrover-logo -m matrix.txt
\end{verbbox}
\fbox{\theverbbox[t]}

\subsection{Using random shuffles as control}
In case you only have a set of signal sequences and you want to use dinucleotide shuffles as control, you have multiple possibilities.
\subsubsection{Generating a set of dinucleotide shuffled sequences}
This software package comes with a C++ implementation of the Altschul-Erkison dinucleotide shuffling algorithm originally implemented in the Python program \verb|altschulEriksonDinuclShuffle.py| by P. Clote, Oct 2003.
It is available in the program \dinuclshuffle{}.
To generate for each sequence contained within a set of signal sequence a shuffled sequence of equal length and equal dinucleotide counts, use the following command:\\
\begin{verbbox}
discrover-shuffle signal.fa > control.fa
\end{verbbox}
\fbox{\theverbbox[t]}

Note that for purposes of reproducibility, you can provide a seed (\verb|--salt|) for the random shuffle generation.

\subsubsection{Automatic generation of random dinucleotide shuffles}
You can also use the same shuffling routines by providing only a set of signal sequences to either \plasma{} or \discrover{}, as the following two commands exemplify.
For \plasma{}\\
\begin{verbbox}
plasma signal.fa -m 8
\end{verbbox}
\fbox{\theverbbox[t]}\\
and for \discrover{}\\
\begin{verbbox}
discrover signal.fa -m 8
\end{verbbox}
\fbox{\theverbbox[t]}

In both cases, dinucleotide shuffles will be generated using the same routines as in \dinuclshuffle{}, and subsequently motif discovery will be executed for motifs that are discriminative between \verb|signal.fa| and the randomly generated set of dinucleotide shuffles.

Note that also in this case for the purpose of reproducibility, you can provide a seed (\verb|--salt|) for the random shuffle generation.


\section{Recipes}
The following section gives some recipes for the analysis of RNA-binding protein data and of ChIP-Seq data.
While the recipes are written for \discrover{}, they can be easily modified to work for \plasma{}:
simply leave out the options \verb|--multiple| and \verb|--wiggle|, which are not supported by \plasma{}.

The recipes use advanced syntax to specify motifs, scores, and sequences in contrasts.
The syntax is explained in detail in section~\ref{section:specification}.

\subsection{Analysing RNA-binding proteins}
First, let us perform a simple analysis and find a motif that is differential between a set of PAR-CLIP signal sequences \verb|rbm10.signal.fa| and a set of matching control sequences \verb|rbm10.control.fa|.
\\
\begin{verbbox}
discrover \
RBM10:rbm10.signal.fa \
control:rbm10.control.fa \
-m RBM10:5-10x3 --multiple --wiggle 1
\end{verbbox}
\fbox{\theverbbox[t]}

\subsubsection{Using repeat experiments}
If a repeat experiment is available, we can define two contrasts \verb|Exp1| and \verb|Exp2|.
Notice how we modify the sequence set definitions to group the sets of sequences into these two contrasts.
\\
\begin{verbbox}
discrover \
RBM10:Exp1:rbm10-experiment1.signal.fa \
control:Exp1:rbm10-experiment1.control.fa \
RBM10:Exp2:rbm10-experiment2.signal.fa \
control:Exp2:rbm10-experiment2.control.fa \
-m RBM10:5-10x3 --multiple --wiggle 1
\end{verbbox}
\fbox{\theverbbox[t]}


\subsection{Analysing ChIP-Seq data}
The first ChIP-Seq recipe finds multiple motifs in a pair of sequence sets, a set of Nanog ChIP-Seq signal sequences \verb|nanog-chen.signal.fa| and a set of matching control sequences \verb|nanog-chen.control.fa|.
It uses the best three seeds for each length between 5 and 16 nucleotides.
We also consider one-nucleotide-shifted variants of each seed.
Thus, in total $11\times3\times3=108$ HMMs will be initialized and individually trained, before the \verb|--multiple| mode begins to build HMMs consisting of multiple motifs.
\\
\begin{verbbox}
discrover -r \
Nanog:nanog-chen.signal.fa \
control:nanog-chen.control.fa \
-m Nanog:5-16x3 --multiple --wiggle 1
\end{verbbox}
\fbox{\theverbbox[t]}

\subsubsection{Using repeat experiments}
In the following recipe we make use of a set of repeat experiments, a set of Nanog ChIP-Seq signal sequences \verb|nanog-marson.signal.fa| produced by a different group, as well as matching control sequences \verb|nanog-marson.control.fa|.
Notice, that the sequence sets used above are grouped into the contrast \verb|Nanog-Chen|, while the second of sequence sets are grouped into the contrast \verb|Nanog-Marson|.
\\
\begin{verbbox}
discrover -r \
Nanog:Nanog-Chen:nanog-chen.signal.fa \
control:Nanog-Chen:nanog-chen.control.fa \
Nanog:Nanog-Marson:nanog-marson.signal.fa \
control:Nanog-Marson:nanog-marson.control.fa \
-m Nanog:5-16x3 --multiple --wiggle 1
\end{verbbox}
\fbox{\theverbbox[t]}

\subsubsection{Motifs differential in one contrast but not in others}
Aside from repeat experiments, we can also search for motifs that are differential in one contrast (or in multiple contrasts if data of repeat experiments are available) but that are not differential in other contrasts.
The following recipe demonstrates this.

In addition to the two contrasts defined in the preceding recipe, it defines two further contrasts of sets of sequences coming from ChIP-Seq experiments for another protein.
Notice how the argument for the \verb|--score| command line switch defines the prior two contrasts to act positively, while the latter two contrasts are taken to act negatively.
\\
\begin{verbbox}
discrover -r \
Nanog:Nanog_Chen:nanog-chen.signal.fa \
control:Nanog_Chen:nanog-chen.control.fa \
Nanog:Nanog_Marson:nanog-marson.signal.fa \
control:Nanog_Marson:nanog-marson.control.fa
Oct4:Oct4_Chen:oct4-chen.signal.fa \
control:Oct4_Chen:oct4-chen.control.fa \
Oct4:Oct4_Marson:oct4-marson.signal.fa \
control:Oct4_Marson:oct4-marson.control.fa \
-m Nanog:5-16x3 --multiple --wiggle 1 --score \
Nanog:+Nanog_Chen+Nanog_Marson-Oct4_Chen-Oct4_Marson:mi
\end{verbbox}
\fbox{\theverbbox[t]}

Notice also that the preceding recipe mentions a motif named \verb|Oct4| to be enriched in the Oct4 signal sequence sets, but does not actually define or train it!
A slight modification of this recipe (replace the motif \verb|Nanog| by \verb|Oct4| in the arguments of \verb|-m| and \verb|--score|, and swap the signs in front of the contrasts in the argument of \verb|--score|) would search for motifs that are differential in the Oct4 contrasts but not differential in the Nanog contrasts.


\section{Usage}
There are two main programs that are part of this package, a seed discovery program, \plasma{}, and one for optimizing binding site HMMs, \discrover{}.
The seed discovery program \plasma{} can be used independently, but it is also used by \discrover{} for automatic seed discovery.
The command line options that drive both programs are intentionally similar.

Below, in section~\ref{subsection:running-plasma}, we will first describe how to run \plasma{} independently, before section~\ref{subsection:running-discrover} describes how to use \discrover{}.

\subsection{Seed discovery with \plasma{}}
\label{subsection:running-plasma}
\subsubsection{Getting help on command line options}
An overview of the available command line options is available with the switch \verb|-h| or its long variant \verb|--help|.
See section~\ref{appendix:plasma-cli-help} for a listing of the output of this command with \plasma{} in version \version{}.

\paragraph{Man page for \plasma{}}
A UNIX man page has been written for \plasma{}.
You can read it by executing \verb|man plasma|.

\subsubsection{Determining the version}
The command line switch \verb|--version| displays the version of \plasma{}.
If additionally the switch \verb|-v| or its long variant \verb|--verbose| is given, then additionally the full SHA1-string of the version is displayed.

\subsubsection{Seed discovery algorithms}
\label{section:plasma-MCMC}
There are multiple algorithms implemented in \plasma{} that allow you to find IUPAC regular expression type seeds.
You can select between them with the switch \verb|--algo|.
It takes one argument, which specifies which algorithm to use.
The following algorithms are provided:
\begin{description}
  \item[Plasma]
    The default and fast algorithm.
    It can be used with the command line switch \verb|--algo plasma|.
    It uses a progressive algorithm that maintains a pool of candidates.
  \item[DREME]
    DREME \cite{Bailey2011}, which is part of the MEME motif discovery package, can be used for motif finding if it is installed on the system.
    It can be used by specifying \verb|--algo dreme|.
  \item[Monte-Carlo Markov Chain optimization]
    \plasma{} finds seeds with Monte-Carlo Markov chain (MCMC) sampling with the command line switch \verb|--algo mcmc|.
    It uses parallel tempering~\cite{Earl2005}, which is also known as replica-exchange, to increase the efficiency of the sampling.
    By default, six parallel chains are used.
\end{description}

\subsubsection{Specifying input files, motifs, and scores}
Please refer to the section~\ref{section:specification} which details the syntax to specify input files, motifs, and scores for \discrover{}.
The syntax for \plasma{} is nearly identical, except for two small differences in the syntax of motif specifications (section~\ref{section:motif-specification}):
\begin{itemize}
  \item
    \plasma{} does not take \Q{insert} terms, and
  \item
    PWMs can not be used as motif specifications.
\end{itemize}


\subsection{Binding site HMM optimization with \discrover{}}
\label{subsection:running-discrover}
\subsubsection{Getting help on command line options}
An overview of the available command line options is available with the switch \verb|-h| or its long variant \verb|--help|.
See section~\ref{appendix:discrover-cli-help} for a listing of the output of this command with \discrover{} in version \version{}.

Note that by default not all options are shown.
To display more comprehensive usage options please specify \verb|-hv| or \verb|--help --verbose|.

In addition, some advanced options are only shown when the option \verb|-V| or \verb|--noisy| is used.
Thus, to display all options please specify \verb|-hV| or their long variants \verb|--help --noisy|.

\paragraph{Man page for \discrover{}}
A UNIX man page has been written for \discrover{}.
You can read it by executing \verb|man discrover|.

\subsubsection{Determining the version}
The command line switch \verb|--version| displays the version of \discrover{}.
If additionally the switch \verb|-v| or its long variant \verb|--verbose| is given, then additionally the full SHA1-string of the version is displayed.

\subsubsection{Hybrid learning and alternative learning modes}
\label{section:hybrid-learning}
By default, \discrover{} uses the hybrid learning mode that is described in the paper.
Here, we refer to the emission parameters of the motif states as signal parameters.
All other parameters, including the transition parameters and the emission parameters of the background state, are referred to as context parameters.
In short, hybrid learning optimizes the emission parameters of the motifs so as to increase the chosen objective function, while the context parameters are optimized for the likelihood.

Aside from the hybrid learning mode, \discrover{} offers two other modes, determined by the \verb|--bglearn| command line switch.
With \verb|--bglearn fixed| the context parameters are left in their initial state and not optimized, while with \verb|--bglearn gradient| the context parameters are optimized for the same chosen objective function.
Although the hybrid learning mode is used by default, you may explicitly specify its usage with \verb|--bglearn em|.

This option is only shown in the full command line help, available with \verb|discrover -hV|.

\subsubsection{Gradient and Monte-Carlo Markov chain optimization}
Gradient optimization is the default learning mode for \discrover{}.
Aside from gradient learning, it is also possible to use Monte-Carlo Markov chain optimization to learn HMM parameters.
This is done with the command line switch \verb|--sampling|.
Like the MCMC optimization mode of \plasma{} (see section~\ref{section:plasma-MCMC}), also that of \discrover{} uses parallel tempering.

The MCMC learning mode of \discrover{} does not support multiple objective learning as in the hybrid learning mode.
For this reason, MCMC learning in \discrover{} optimizes all parameters, including the context parameters\footnote{Transition parameters and the emission parameters of the background state, see section~\ref{section:hybrid-learning}.}, for the chosen objective function.

There are several options available to determine the way \discrover{} performs MCMC\@.
For these, please refer to the \emph{MCMC optimization options} section of the extended command line help, available via \verb|discrover --help --verbose|.

\subsection{Specifying input files, motifs, and scores}
\label{section:specification}
The following section details the syntax to specify input files, motifs, and scores for \discrover{}.

\newcommand{\Item}[1]{\item[#1] \hfill \\}

\subsubsection{FASTA file specification}
FASTA file specifications consist of up to three colon-separated terms, with the following syntax:

\begin{verbbox}
[names:[contrast:]]path
\end{verbbox}
\fbox{\theverbbox[t]}

The square brackets indicate that \Q{names} and \Q{contrast} are optional.
The meaning of \Q{names}, \Q{contrast}, and \Q{path} are as follows:

\begin{description}
 \Item{\Q{names}}
   comma-separated list of names of motifs enriched in this file, or \Q{control}
 \Item{\Q{contrast}}
   name of a contrast this file belongs to
 \Item{\Q{path}}
   path of FASTA file with sequences
\end{description}

All FASTA files for which no \Q{contrast} is given are associated with a common, automatically named contrast.

Using contrasts, FASTA files can be grouped into contrasts on which score specifications can act.
If \Q{path} contains at least one colon, please prepend colons to disambiguate.

FASTA files tagged \Q{control} are treated slightly differently:
they are excluded from the set of sequences on which parameter re-estimation is performed (motif occurrence priors and background emissions will thus not be learned from these).
Note: shuffled sequences are implicitly tagged \Q{control}.

\subsubsection{Motif specification}
\label{section:motif-specification}
Motif specifications consist of up to three colon-separated terms, with the following syntax:

\begin{verbbox}
[name:[insert:]]motifdef
\end{verbbox}
\fbox{\theverbbox[t]}

The square brackets indicate that \Q{name} and \Q{insert} are optional.
The meaning of \Q{name}, \Q{insert}, and \Q{motifdef} are as follows:

\begin{description}
  \Item{\Q{name}}
    unique name for the motif; may not be \Q{control}
  \Item{\Q{insert}}
    comma-separated list of positions after which to add an insertion state; positions are 1-indexed; $1 \leq \text{position} < \text{motif length}$ must hold for each position
  \Item{\Q{motifdef}}
    motif definition; see below
\end{description}

If \Q{name} is not given, a name of the form \Q{motifN} is generated, where \Q{N} are numbers increasing from 0.

There are three ways in which a \Q{motifdef} can be given:

\begin{description}
  \Item{IUPAC regular expression}
    Using the IUPAC code for nucleic acids.
  \Item{Length specification}
    Motifs of the indicated lengths are sought.
    A length specification is a comma separated list of length ranges, where a length range is either a single length, or an expression of the form \Q{M-N} to indicate lengths \Q{M} up to \Q{N}.
    A length specification also allows to specify a multiplicity separated by an \Q{x}.
    Thus examples are \Q{8} for just length 8, \Q{5-7} for lengths 5, 6, and 7, \Q{5-8x2} for two motifs of lengths 5 to 8, \Q{5-8,10x3} for three motifs of lengths 5, 6, 7, 8, and 10.
  \Item{Path to a file with a position weight matrix (PWM)}
    If the path contains at least one colon, please prepend colons to disambiguate.
\end{description}

By default, insert states given by an \Q{insert} specification do not have self-transition edges.
You can add self-transition edges to the insert states by using the command line switch \Q{--selftrans}.
Please refer to figure~\ref{fig:insert-states} for an illustration.

\begin{figure}
  \centering
  \ignore{%
    \TikzEnv{%
      \SimpleCore{}
    }
    \TikzEnv{%
      \SimpleCore{}
      \Insertions{}
    }
  }
  \TikzEnv{%
    \SimpleCore{}
    \begin{scope}[color=blue]
      \Insertions{}
    \end{scope}
    \begin{scope}[color=red]
      \InsertionsSelf{}
    \end{scope}
  }
  \ignore{%
    \TikzEnv{%
      \ExtendedCore{}
      \Insertions{}
      \InsertionsSelf{}
      \Deletions{}
    }
    \TikzEnv{%
      \ExtendedCore{}
      \Insertions{}
      \InsertionsSelf{}
      \Deletions{}
      \InsertionToDeletionEdges{}
    }
  }
  \caption{%
    \label{fig:insert-states}
    Insert states.
    By default, \discrover{} uses motif topologies as indicated by the black states.
    $B$ denotes the background state, and states $M_1$--$M_8$ denote states 1--8 of a motif of length 8.
    Such a topology is used for example with the motif specifiction \Q{-m motif:8}.
    Insert states $I_1$--$I_7$ (shown in blue) are added with a motif specification such as \Q{-m motif:1,2,3,4,5,6,7:8}.
    If self-transitions (shown in red) are wanted the command line switch \Q{--selftrans} has to be used as in \Q{-m motif:1,2,3,4,5,6,7:8 --selftrans}.
  }
\end{figure}

\subsubsection{Score specification}
Score specifications consist of up to three colon-separated terms, with the following syntax:

\begin{verbbox}
[motif:[contrasts:]]measure
\end{verbbox}
\fbox{\theverbbox[t]}

The square brackets indicate that \Q{motif} and \Q{contrasts} are optional.
The meaning of \Q{motif}, \Q{contrasts}, and \Q{measure} are as follows:

\begin{description}
  \Item{\Q{motif}}
    name of the motif to optimize
  \Item{\Q{contrasts}}
    contrast specification; see below
  \Item{\Q{measure}}
    significance measure to use
\end{description}

If \Q{name} is not given given, all motifs (that are not named in any other score specification) are optimized with this score specification.
It is an error when more than one score specification is given for any motif.

The contrast specification \Q{contrasts} is a set of contrast names prefixed by \Q{+} or \Q{-} to indicate whether the score on the contrast is to count positively or negatively.
The first contrast name, if lacking a \Q{+} or \Q{-}, is taken to count positively.
If \Q{contrasts} is not given, the score definition uses all contrasts, counting each one positively.

The following significance measures are available:

\begin{description}
  \Item{\Q{none}}
    Do not perform learning
  \Item{\Q{bw}}
    Likelihood maximization with the Baum-Welch algorithm; this measure is not discriminative
  \Item{\Q{viterbi}}
    Viterbi learning; this measure is not discriminative
  \Item{\Q{mi}}
    Mutual information of condition and motif occurrence (MICO)
  \Item{\Q{ri}}
    Rank mutual information
  \Item{\Q{mmie}}
    Maximum mutual information estimation (MMIE), also known as posterior classification probability
  \Item{\Q{mcc}}
    Matthews correlation coefficient
  \Item{\Q{dlogl}}
    Log-likelihood difference, like in the motif discovery method DME, see doi: 10.1073/pnas.0406123102
  \Item{\Q{dfreq}}
    Difference of frequency of sequences with motif occurrences, similar to the motif discovery methods DIPS and DECOD, see doi: 10.1093/bioinformatics/btl227 and 10.1093/bioinformatics/btr412
\end{description}

\subsection{Output files}
The output files produced by \discrover{} comprise:

\begin{description}
  \Item{\Q{.hmm}}
    Parameter of the trained HMM\@.
    May be loaded later with \Q{--learn}.
  \Item{\Q{.summary}}
    Summary information with number of occurrences of the motifs in each sample, and various generative and discriminative statistics.
  \Item{\Q{.viterbi}}
    FASTA sequences annotated with the Viterbi path, and sequence level statistics.
  \Item{\Q{.bed}}
    BED file of coordinates of matches to the motifs in all sequences.
  \Item{\Q{.table}}
    Coordinates and sequences of matches to the motifs in all sequences (extends the \Q{.bed} output file).
  \Item{\Q{.pdf} / \Q{.png}}
    Sequence logos of the found motifs in PDF / PNG format.
\end{description}

Note that, depending on the argument of \Q{--compress}, the \Q{.viterbi}, \Q{.bed}, and \Q{.table} files may be compressed, and require decompression for inspection.


\section{Appendix}
\subsection{Command line help of \plasma{}}
\label{appendix:plasma-cli-help}
{%
  \footnotesize
  \verbatiminput{../plasma-cli-help.txt}
}

\subsection{Command line help of \discrover{}}
\label{appendix:discrover-cli-help}
{%
  \footnotesize
  \verbatiminput{../discrover-cli-help.txt}
}



\nocite{*}
\phantomsection
\addcontentsline{toc}{section}{References}
\bibliographystyle{plainnat}
\bibliography{references}


\end{document}
